### PR TITLE
gpu: support strict gpu share with hami

### PR DESCRIPTION
**Pull request by ZiMengSheng**
### Ⅰ. Describe what this PR does

Support GPU Share With HAMi

### V. Checklist

- [ ] I have written necessary docs and comments
- [ ] I have added necessary unit tests and integration tests
- [ ] All checks passed in `make test`

### CODE REVIEW

File: pkg/koordlet/runtimehooks/hooks/gpu/gpu.go [hunk 1]
```diff
@@ -70,5 +71,30 @@ func (p *gpuPlugin) InjectContainerGPUEnv(proto protocol.HooksProtocol) error {
...
+		return fmt.Errorf("gpu core not found in gpu resource")
+	}
+	gpuMemoryRatio, ok := gpuResources[ext.ResourceGPUMemoryRatio]
+	if !ok {
+		return fmt.Errorf("gpu memory ratio not found in gpu resource")
+	}
+	if gpuMemoryRatio.Value() < 100 {
+		containerCtx.Response.AddContainerEnvs["CUDA_DEVICE_MEMORY_LIMIT"] = fmt.Sprintf("%d", gpuMemory.Value())
+		containerCtx.Response.AddContainerEnvs["CUDA_DEVICE_SM_LIMIT"] = fmt.Sprintf("%d", gpuCore.Value())
+		containerCtx.Response.AddContainerEnvs["LD_PRELOAD"] = "/libvgpu.so"
```
Comment (saintube) [inline pkg/koordlet/runtimehooks/hooks/gpu/gpu.go:91]: Shall we add a feature-gate for HAMI?

Comment (ZiMengSheng) [reply pkg/koordlet/runtimehooks/hooks/gpu/gpu.go:91]: Yep，maybe a label is better

File: pkg/koordlet/runtimehooks/protocol/container_context.go [hunk 2]
```diff
@@ -181,8 +181,9 @@ func (c *ContainerRequest) FromReconciler(podMeta *statesinformer.PodMeta, conta
 }
 
 type ContainerResponse struct {
-	Resources        Resources
-	AddContainerEnvs map[string]string
+	Resources          Resources
+	AddContainerEnvs   map[string]string
+	AddContainerMounts []*api.Mount
```
Comment (saintube) [inline pkg/koordlet/runtimehooks/protocol/container_context.go:186]: We'd better avoid referring to NRI API directly in the koordinator protocol. Consider making a new Mount struct and we can transform the internal protocol to the NRI protocol or the CRI protocol.

File: apis/extension/device_share.go [hunk 3]
```diff
@@ -58,6 +58,8 @@ const (
 	LabelGPUModel                   string = NodeDomainPrefix + "/gpu-model"
 	LabelGPUDriverVersion           string = NodeDomainPrefix + "/gpu-driver-version"
 	LabelSecondaryDeviceWellPlanned string = NodeDomainPrefix + "/secondary-device-well-planned"
+
+	LabelGPUIsolationProvider = "koordinator.sh/gpu-isolation-provider"
```
Comment (saintube) [inline apis/extension/device_share.go:62]: `DomainPrefix + "/gpu-isolation-provider"` is better

File: pkg/koordlet/util/system/config.go [hunk 4]
```diff
@@ -77,30 +78,32 @@ func InitSupportConfigs() {
...
-		DefaultRuntimeType:    "containerd",
+		ProcRootDir:                  "/proc/",
+		SysRootDir:                   "/host-sys/",
+		SysFSRootDir:                 "/host-sys-fs/",
+		VarRunRootDir:                "/host-var-run/",
+		VarLibKubeletRootDir:         "/var/lib/kubelet/",
+		RunRootDir:                   "/host-run/",
+		RuntimeHooksConfigDir:        "/host-etc-hookserver/",
+		DefaultRuntimeType:           "containerd",
+		HAMICoreLibraryDirectoryPath: "/data/bin/libvgpu.so",
```
Comment (carlory) [inline pkg/koordlet/util/system/config.go:106]: Why the default path is `/data/bin/libvgpu.so`? According to [hami chart](https://github.com/Project-HAMi/HAMi/blob/master/charts/hami/templates/device-plugin/daemonsetnvidia.yaml), the default host path is `/usr/local/vgpu/libvgpu.so`.

Comment (ZiMengSheng) [reply pkg/koordlet/util/system/config.go:106]: OK, I will fix it

File: pkg/koordlet/runtimehooks/hooks/gpu/gpu.go [hunk 5]
```diff
@@ -70,5 +71,31 @@ func (p *gpuPlugin) InjectContainerGPUEnv(proto protocol.HooksProtocol) error {
 		containerCtx.Response.AddContainerEnvs = make(map[string]string)
 	}
 	containerCtx.Response.AddContainerEnvs[GpuAllocEnv] = strings.Join(gpuIDs, ",")
+	if containerReq.PodLabels[ext.LabelGPUIsolationProvider] == string(ext.GPUIsolationProviderHAMICore) {
+		gpuResources := devices[0].Resources
```
Comment (saintube) [inline pkg/koordlet/runtimehooks/hooks/gpu/gpu.go:75]: Shall we ensure the target GPU info is on the first device?

File: pkg/koordlet/runtimehooks/protocol/container_context.go [hunk 6]
```diff
@@ -288,6 +289,17 @@ func (c *ContainerContext) NriDone(executor resourceexecutor.ResourceUpdateExecu
 		}
 	}
 
+	if len(c.Response.AddContainerMounts) != 0 {
```
Comment (saintube) [inline pkg/koordlet/runtimehooks/protocol/container_context.go:292]: no need to check the length for the for-range

File: pkg/koordlet/runtimehooks/hooks/gpu/gpu.go (hunk 5 above)
Comment (ZiMengSheng) [reply pkg/koordlet/runtimehooks/hooks/gpu/gpu.go:75]: currently, we assure all gpu device has the same allocation result

**Review by saintube — APPROVED** [review]
/lgtm

**Pull request merged**